src/nnet3/nnet-cudnn-simple-component.cc

```diff
@@ -0,0 +1,798 @@
+                                 );
+}
+
+Component *CuDNN3DConvolutionComponent::Copy() const {
+  CuDNN3DConvolutionComponent *ans = new CuDNN3DConvolutionComponent(*this);
+  return ans;
+}
+
+void CuDNN3DConvolutionComponent::Read(std::istream &is, bool binary) {
+  ReadUpdatableCommon(is, binary);  // Read opening tag and learning rate.
+  ExpectToken(is, binary, "<InputXDim>");
+  ReadBasicType(is, binary, &input_x_dim_);
+  ExpectToken(is, binary, "<InputYDim>");
+  ReadBasicType(is, binary, &input_y_dim_);
+  ExpectToken(is, binary, "<InputZDim>");
+  ReadBasicType(is, binary, &input_z_dim_);
+  ExpectToken(is, binary, "<NumFilters>");
+  ReadBasicType(is, binary, &num_filters_);
+  int32 filter_dims[kConvolutionDimension_];
+  ExpectToken(is, binary, "<FilterXDim>");
+  ReadBasicType(is, binary, &filter_dims[0]);
+  ExpectToken(is, binary, "<FilterYDim>");
+  ReadBasicType(is, binary, &filter_dims[1]);
+  ExpectToken(is, binary, "<FilterZDim>");
+  ReadBasicType(is, binary, &filter_dims[2]);
+  int32 padding[kConvolutionDimension_];
+  ExpectToken(is, binary, "<FilterXPadding>");
+  ReadBasicType(is, binary, &padding[0]);
+  ExpectToken(is, binary, "<FilterYPadding>");
+  ReadBasicType(is, binary, &padding[1]);
+  ExpectToken(is, binary, "<FilterZPadding>");
+  ReadBasicType(is, binary, &padding[2]);
+  int32 strides[kConvolutionDimension_];
+  ExpectToken(is, binary, "<FilterXStride>");
+  ReadBasicType(is, binary, &strides[0]);
+  ExpectToken(is, binary, "<FilterYStride>");
+  ReadBasicType(is, binary, &strides[1]);
+  ExpectToken(is, binary, "<FilterZStride>");
+  ReadBasicType(is, binary, &strides[2]);
+  int32 upscales[kConvolutionDimension_];
+  ExpectToken(is, binary, "<FilterXUpscale>");
+  ReadBasicType(is, binary, &upscales[0]);
+  ExpectToken(is, binary, "<FilterYUpscale>");
+  ReadBasicType(is, binary, &upscales[1]);
+  ExpectToken(is, binary, "<FilterZUpscale>");
+  ReadBasicType(is, binary, &upscales[2]);
+  ExpectToken(is, binary, "<InputVectorization>");
+  int32 input_vectorization;
+  ReadBasicType(is, binary, &input_vectorization);
+  input_vectorization_ = static_cast<TensorVectorizationType>(input_vectorization);
+  ExpectToken(is, binary, "<FilterParams>");
+  filter_params_.Read(is, binary);
+  ExpectToken(is, binary, "<BiasParams>");
+  bias_params_.Read(is, binary);
+  ExpectToken(is, binary, "<IsGradient>");
+  ReadBasicType(is, binary, &is_gradient_);
+  ExpectToken(is, binary, "</CuDNN3DConvolutionComponent>");
+
+  CUDNN_SAFE_CALL(cudnnSetFilterNdDescriptor(filter_desc_,
+                                             cudnn::GetDataType(),
+                                             kConvolutionDimension_,
+                                             filter_dims
+                                             ));
+  // N x C x D x H x W
+  // TODO: torch uses {1, num_filters, 1, 1}
+  // Understand why. Is this a mistake?
+  int32 bias_dims[] = {1, num_filters_, 1, 1};
```

Comment:
This is a little fishy. I think I believed that bias_dims should be `{1, num_filters_, 1, 1, 1}`, as each filter has its own bias scalar, and each filter is 3 dimensional, not 2 dimensional.